### ISSUE TITLE
Class attribute inheritance deviation

### ISSUE THREAD

**Issue by maximlt** [holoviz/param#718]
In a hierarchy of classes having class attributes, updating the class attribute in a super class is seen by the subclasses. Below `B.p` resolves to the updated value of `p` on `A`.

```python
class A:
    p = 1

class B(A):
    pass

print(B.p)
# 1

A.p = 2

print(B.p)
# 2
```

This is not how Param behaves, updating `p` on `A` has no effect on the subclasses. Not sure to which extent this is a bug.

```python
import param

class A(param.Parameterized):
    p = param.Parameter(1)
    
class B(A):
    p = param.Parameter()

print(B.p)
# 1

A.p = 2

print(B.p)
# (Different behavior)
# 1
```

**Comment by jbednar** [comment]
Interesting question. Param does behave like standard class attributes when the Parameter is not defined on the subclass:

```python
import param

class A(param.Parameterized):
    p = param.Parameter(1)
    
class B(A):
    pass

print(B.p)
# 1

A.p = 2

print(B.p)
# 2
```

But when a Parameter is defined on the subclass, the new definition takes precedence, and default value inheritance happens only at instantiation time, not runtime.

We can certainly clarify this in the docs with an explicit statement that default value inheritance only occurs when a subclass does not redefine the parameter. But I'm open to having the subclass dynamically inherit the _current_ default value from the superclass once we have [sentinel support](https://github.com/holoviz/param/pull/605) to record whether the default value has ever been set in the subclass.

**Comment by maximlt** [comment]
Just observed this is still the behavior in Param 2.0 after the sentinel support has been merged. I briefly tried to see how that could be made to work but quickly faced some issues. That was a very very quick attempt and there must be a better way to approach it, still sharing notes:

- Making it "work" with something like: skip setting `default` in `__param_inheritance`, and update `Parameter.__get__` to walk up the class hierarchy in search for the Parameter with a defined default.
- Issues: Some Parameter attributes aren't static but trigger computations at the Parameter level that need `default` to be defined

I don't know, it may be that Param has become too dynamic for its own good, with "vertical" (parameter attribute inheritance) and "horizontal" (parameter attribute depending on other attributes) dimensions. Something to keep in mind and check whether it doesn't prevent more innovation (typing!).

**Comment by jbednar** [comment]
Yep, now's the time to investigate this again, now that sentinel support makes it possible. I agree those will be complications!
